OverspeedEventHandler enhancement to use attributes

Abyss777 opened the pull request:
I've changed OverspeedEventHandler to use speed limit from attributes. Attribute name is `speedLimit`
Attribute lookup starts from Device -> Groups -> Server and Device has higher priority

Because lookup groups is expensive, I've added `event.overspeed.groupsEnabled` config to enable it.
There is two idea: rise event on every "overspeeding" point (by default) or only once during period.
`event.overspeed.riseOnce` enables it.

Also I've removed Device.status from DB. It is really small commit to separate it.

Tested for a two days. Works OK.

tananaev commented on src/org/traccar/events/OverspeedEventHandler.java:
```diff
@@ -50,8 +55,48 @@ public OverspeedEventHandler() {
+            }
+        }
+        speedLimit = UnitsConverter.knotsFromKph(speedLimit);
+        if (speedLimit != 0 && speed > speedLimit && oldSpeed <= speedLimit) {
```
Have you tested the logic? It seems like if `riseOnce` (by the way, you should probably rename it to something like `repeat`) is set to true, `oldSpeed` will be always 0 and you will always get an event. Am I missing something?

Abyss777 replied:
Yes, I have.
As you suggested in email discussion, by default it should repeat events on every "overspeeding" point. `riseOnce` disables this behavior, "Once" is right word. 
Actually if `riseOnce` is false `oldSpeed` will be always zero, and third part of compare will be always true. (`oldSpeed <= speedLimit`)
Because of that whole compare will be true only if `speed > speedLimit` and it is default behavior.

tananaev replied:
Wait, so we request events from the database on every position report? When do we remove events from the database?

Abyss777 replied:
Actually, yes.  We do it to avoid too frequent events. We never remove events from DB.

tananaev replied:
I guess I missed it before. We should get rid of this behaviour as soon as possible. What if we have millions of events in the database? We should never request anything from the database for every position.

Abyss777 replied:
You are right. I'll do that. Events can be filtered later in reports ore somewhere else.

tananaev replied:
Thanks. It applies for all events. I guess we can have another event filtering handler (which can cache old events or just type and timestamp or something).

Abyss777 replied:
But... we still requesting server object from DB...

tananaev replied:
Good point. Maybe we should cache it as well, but server table always has only one record, so it shouldn't be as big of a problem as requesting events.

Abyss777 replied:
Actually `Server` is cached in `PermissionManager` but used only there and never updated.
For example updating `readOnly` via API won't affect until sever restart.
I'll organize it too.

tananaev replied:
Great, thanks again.

tananaev commented on src/org/traccar/events/OverspeedEventHandler.java:
```diff
@@ -50,8 +55,48 @@ public OverspeedEventHandler() {
+        if (device.getAttributes().containsKey(ATTRIBUTE_SPEED_LIMIT)) {
+            speedLimit = Double.parseDouble((String) device.getAttributes().get(ATTRIBUTE_SPEED_LIMIT));
+        }
+        if (speedLimit == 0 && checkGroups) {
```
Should we move attribute lookup logic somewhere else. It will be reused for other attributes as well, right?

Abyss777 replied:
Yes, It can be used elsewhere . What is right place to it? `DeviceManager`?

tananaev replied:
`DeviceManager` might be a good place, or just another utility class.

Abyss777 commented:
My mistake, Server object was actually being updated with `refresh()` but I remember case that it did not affect on `readOnly` .  Anyway  I added proxy functions to retrieve and update cached object. And made it `volatile` to work correctly from different threads.

Detached `lookupAttribute` to function and moved to `DeviceManager`. There is side effect. Because functions does not know type of looking attribute it stops looking just when find it. And if "speedLimit" is zero handler will always skip it. That means users can made exclusions for devices or groups just setting "speedLimit" to zero.

Removed all DB requests during analyzing positions.

Tested on only one real trip. Looks working OK.

tananaev commented on debug.xml:
```diff
@@ -41,10 +41,11 @@
-    <entry key='event.suppressRepeated'>60</entry>
     <entry key='event.overspeedHandler'>true</entry>
-    <entry key='event.globalSpeedLimit'>90</entry>
+    <entry key='event.overspeed.riseOnce'>true</entry>
```
Can you please rename parameter to `repeat`. Rise once sounds a bit confusing.

Abyss777 replied:
The meaning is opposite. `notRepeat` is fit?

tananaev replied:
We can just inverse the logic and set the name to "repeat". I guess
"notRepeat" is also OK.

Abyss777 replied:
Inversing the logic will inverse default behavior. If we omit parameter it is assumed as "false". 
Should I inverse default behavior?

tananaev replied:
I guess we can. Default doesn't really matter that much because you can
always explicitly specify what you want.

tananaev commented:
Just one small comment. Otherwise looks good. Assuming that it works :)

Abyss777 commented:
I've renamed. If need inversing, I'll do it a bit later.

tananaev commented:
Merged, thanks.
